### PR TITLE
apply patches in issues #129 and #124

### LINKED ISSUE #121: metacharacters produce FN validation
```perl
sub is_valid_p {
    my ( $self, $p ) = @_;
    croak "unspecified p" if !defined $p;
    return ( grep {/^$p$/i} qw/ none reject quarantine / ) ? 1 : 0;
}
```
can be spoofed with `"v=DMARC1; p=.*"`

not that that is of any real value, but the idea of modding the pattern /^$x$/ to /^\Q$x\E$/ is what's intended here.

### LINKED ISSUE #123: Mail::DMARC::Report::Send prepends a relative directory to search path
There is a stray `use lib` with a relative directory in `Mail::DMARC::Report::Send`. This prepends this directory, which is then searched first relative to the current directory. This is a problem similar to CVE-2016-1238.

### LINKED ISSUE #124: dmarc_send_reports retries sending forever
Hi,

I've noticed that there are cases where a report is not removed from the database despite failure to send the report. For example, in the case below, the server returns a 554 error. This is a permanent error. Shouldn't dmarc_dend_reports delete this entry from the DB instead of retrying? Perhaps we should examine the $smtp->code in line 395 of dmarc_send_reports.bat and delete records which have resulted in permanent failures, i.e. 5XX codes. I have attached a patch which deals with the error when dealing with DATA command (copied from existing code which dealt with 5XX for recipients).

ID: 44
Domain: isdg.net
rua: mailto:dmarc-rua@isdg.net

getting MX for isdg.net
delivering message to dmarc-rua@isdg.net, via mail.isdg.net
DATA for domain isdg.net report rejected
554 Message Not Accepted by filter.

DBI error: NOT NULL constraint failed: report_error.time
Sending error query called by Mail::DMARC::Report::Store::SQL, 498
INSERT INTO report_error (report_id, error ) VALUES (??)
44, 554 Message Not Accepted by filter.
at C:/Program Files (x86)/Vircom/modusGate/perl/site/lib/Mail/DMARC/Report/Store.pm line 17.

[dmarc_send_reports.patch.txt](https://github.com/msimerson/mail-dmarc/files/1544731/dmarc_send_reports.patch.txt)

### LINKED ISSUE #129: Sending reports in TLS fails for certain domains
Hi, 

I found the following issues when sending reports to certain domains.  

When sending to the MX for rua.agari.com, specifically etl.cp.prod.agari.com, connect_smtp_tls returns a valid smtp object making one think all is good.  There is a bug however as when invoking a starttls, you get a failure.  While the server supports TLS, the handshake fails.  One needs to examine the status code of the starttls before using TLS.  

Secondly, in email STARTTLS is only really used for encryption purposes and the certificate is not necessarily verified.  Self signed certificates do exist (see agari example above).  So the patch includes SSL_VERIFY_NONE when constructing and SMTPS object and also checking the return value.

Third, some servers require a HELO after a STARTTLS (), e.g. mx1.sophos.com.  That is not happening, so I'm issuing a hello after starttls.  A second hello does not hurt.

Finally, the specified timeout kicks in a lot on a lot of live servers.  Have increased to 30 seconds.

See the attached patch:

[starttls.patch.txt](https://github.com/msimerson/mail-dmarc/files/1939310/starttls.patch.txt)

### PR DESCRIPTION
**Pull request by msimerson**
fixes #129 
fixes #124 
fixes #123 
fixes #121

### CODE REVIEW

File: bin/dmarc_send_reports [hunk 1]
```diff
@@ -398,7 +398,21 @@ sub email {
...
+                    'id'        => $rid,
+                    'error'     => 'Report deleted',
+                    'smtp_code' => $smtp->code,
+                });
+                $report->store->delete_report($rid);
+            }
+            else {
+                $report->store->error($rid, $err);
+            };
+        };
```
Comment (marcbradshaw) [inline bin/dmarc_send_reports]: lines 414/415 terminating ; not necessary

**Comment by marcbradshaw** [comment]
A couple of comments made, generally looks good to me.

File: README.md [hunk 2]
```diff
@@ -241,19 +241,19 @@ The SPF result code: none, neutral, pass, fail, softfail, temperror, or permerro
 
 ## Correct
 
-In cases where correctness is ambiguous, options are provided.
+The DMARC spec is lengthy and evolving, making correctness a moving target. In cases where correctness is ambiguous, options are generally provided.
 
 ## Easy to use
 
-Providing an implementation of DMARC that SMTP utilities like SpamAssassin, amavis, and qpsmtpd can consume is expected to increase adoption.
+The effectiveness of DMARC will improve significantly as adoption increases. Proving an implementation of DMARC that SMTP utilities like SpamAssassin, amavis, and qpsmtpd can consume will aid adoption.
```
[the remark this answers is not in the document]
Comment (msimerson) [reply README.md]: Yep. :)

**Comment by djugessur** [comment]
Regarding #129, the fix done is not quite right.    Some servers require a HELO after a STARTTLS *before* they reply with a code.  If you look in the patch provided, I check for the return code after the HELO is sent and that is by design.

**Pull request merged**